### ISSUE TITLE
MIDI Loop

### ISSUE THREAD

**Issue by sinosoidal** [thestk/rtmidi#48]
Hi,

Everything I send thru a MIDI out virtual port is received on the MIDI in virtual port. Is is normal?

Thx,

Nuno

**Comment by garyscavone** [comment]
No responses ... not verified ... closing.

**Comment by sinosoidal** [comment]
If you could not verify, can you tell me more about your setup?

**Comment by garyscavone** [comment]
On OS-X, I made a short test program to open both input and output virtual ports but I did not see data being sent out through the virtual output port showing up as input on my virtual input.  The trick with the virtual ports is getting them correctly connected to the software you want to use them with.

**Comment by sinosoidal** [comment]
I have a desktop app called TKLINK which serves as bridge between a mobile app and a software running on desktop, in this case, TRAKTOR.

By the time I submitted this bug, I was trying to get the feedback of executing certain command on TRAKTOR. I would send a command from the mobile app to TKLINK, that command would be converted to MIDI and sent via virtual midi port to TRAKTOR, which by it’s turn, would reply the new state, again via the same Virtual MIDI port. 

I have only one virtual midi port. Should I have two?

**Comment by garyscavone** [comment]
Your original issue was: "Everything I send thru a MIDI out virtual port is received on the MIDI in virtual port. Is is normal?”

To me, that implies you had both a virtual input and a virtual output port open. Given your last email, I don’t really understand what you are doing or what the problem is.  RtMidi separates MIDI input and output into two separate classes (RtMidiIn and RtMidiOut) and each of those classes can be used to create a virtual port (either a virtual input port via RtMidiIn or a virtual output port via RtMidiOut).  But note that you do not need to create your own virtual port with RtMidiIn or RtMidiOut to connect to an already existing virtual port created by a different application on your system.  For example, if I open the MidiKeys application, which creates its own virtual port, I can simply connect to that port using the RtMidiIn::openPort() function.

Virtual ports are a bit confusing to explain, so I’m not sure I’m helping here.

**Comment by radarsat1** [comment]
Github needs diagrams ;)

sinosoidal, I'm a bit confused by your description.. is TkLink a program you wrote, does it use RtMidi?  I'm not sure this issue could be much better investigated without a more concrete setup and example.. 
Can you provide exactly what functionality is being invoked, what the expected result is, and what you see?

On the other hand, if it's not an on-going problem for you (which I gather from your comment being in the past tense), it would be good to leave the issue closed for now, and open a new one if it continues to bother you..

Apologies in advance if I'm misunderstanding anything.

**Comment by sinosoidal** [comment]
I am in fact confused because its been a while since I had this issue and my memory is not clear about the problem I was having. 

I have just reviewed my code and i’m using two virtual ports. One for input, one for output. I need in fact to try again and see if I can replicate the problem. It might take a while since I’m not focused on something else.

Thanks for your feedback.

With my best regards,

Nuno

**Comment by sinosoidal** [comment]
Just to complement. 

I currently have two mobile apps, LK and TKFX, they both have a bridge, something that announces a bonjour service on a network to which the mobile app will connect to. 

I use RtMidi on both TKLINK and LKBRIDGE. LKBRIDGE opens a midi out port, to receive the network messages and forward them the selected midi port. 

On TKLINK, I needed to receive the midi clock from TRAKTOR so I had to use a midi in port.

TKLINK is the bridge for TKFX, a dedicated traktor effects controller. TRAKTOR is a djing software that can be controlled by midi controller. So, what I do is to send commands to TRAKTOR that will have an impact on the UI. I can’t make TRAKTOR send command for that UI changes in order to be reflected on the app. Here, I get to the point in which  my memory fails and I need to reproduce the problem to get back here! :)

Thanks once again for your feedback.
